autonomi/src/client/high_level/vault/mod.rs

```diff
@@ -78,21 +144,203 @@ impl Client {
         secret_key: &VaultSecretKey,
         content_type: VaultContentType,
     ) -> Result<AttoTokens, VaultError> {
-        let scratch_address = ScratchpadAddress::new(secret_key.public_key());
-        info!("Writing to vault at {scratch_address:?}");
-
-        match self
-            .scratchpad_update(secret_key, content_type, &data)
-            .await
-        {
-            Ok(()) => Ok(AttoTokens::zero()),
-            Err(ScratchpadError::CannotUpdateNewScratchpad) => {
-                let (price, _) = self
-                    .scratchpad_create(secret_key, content_type, &data, payment_option)
-                    .await?;
-                Ok(price)
+        info!("Writing {} bytes to vault ...", data.len());
+        let mut total_cost = AttoTokens::zero();
+        let main_secret_key = MainSecretKey::new(secret_key.clone());
+
+        // scratchpad_derivations ordered by the collection order
+        let (mut cur_free_graphentry_derivation, mut scratchpad_derivations) = self
+            .current_capacity(
+                &main_secret_key,
+                DerivationIndex::from_bytes(VAULT_HEAD_DERIVATION_INDEX),
+            )
+            .await?;
+
+        let content_chunks = chunk_bytes(data);
+
+        info!(
+            "Current capacity is {}, meanwhile requiring {}",
+            scratchpad_derivations.len(),
+            content_chunks.len()
+        );
+
+        // claim more capacity if short of.
+        // Note: as the Scratchpad is `created on use`, hence during the `claim stage`,
+        //       NUM_OF_SCRATCHPADS_PER_GRAPHENTRY to be claimed in one newly created GraphEntry.
+        while scratchpad_derivations.len() < content_chunks.len() {
+            let (new_free_graphentry_derivation, new_scratchpad_derivations, grap_cost) = self
+                .expand_capacity(
+                    &main_secret_key,
+                    &cur_free_graphentry_derivation,
+                    payment_option.clone(),
```

Comment:
Use of payment option in a loop where it could be a receipt. This code would only work if payment option was a wallet as one cannot know in advance what addresses to pay for.

Reply:
as the ScratchPad is now generated on need, it may means only accept Wallet to allow dynamic paying makes more sense.
The current API also doesn't have the facility for the `addresses to pay` to be collected in advance (not a bad thing, as to keep the API simple)

Will prefer to leave the work of `make Vault support pay by receipt` as a separate follow-up PR, if necessary.

Reply:
Agreed, follow up work. Although I wonder if we should keep Vaults auto-upgrading? Instead of returning an error for max capacity reached and offering an API specifically to upgrade? No hidden costs this way but maybe too complex? Stuff to think about...